In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

###### Import the data as panda dataframe and add column names to the data

In [2]:
dataCol = ['user id','movie id','rating','timestamp']
itemCol = ['movie id','movie title','release date','video release date','IMDb URL','unknown','Action',
'Adventure','Animation','Childrens','Comedy','Crime','Documentary','Drama','Fantasy','Film-Noir','Horror',
'Musical','Mystery','Romance ','Sci-Fi','Thriller','War' ,'Western']
userCol = ['user id','age','gender','occupation','zip code']

In [577]:
data = pd.read_csv('ml-100k/u.data', sep='\t', names=dataCol, encoding='latin-1')
item = pd.read_csv('ml-100k/u.item', sep='|', names=itemCol, encoding='latin-1')
user = pd.read_csv('ml-100k/u.user', sep='|', names=userCol, encoding='latin-1')

In [578]:
print data[:10]
print data.shape
print item.shape
print user.shape

   user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596
5      298       474       4  884182806
6      115       265       2  881171488
7      253       465       5  891628467
8      305       451       3  886324817
9        6        86       3  883603013
(100000, 4)
(1682, 24)
(943, 5)


In [16]:
item[:10]

movie id                                        movie title release date  \
0         1                                   Toy Story (1995)  01-Jan-1995   
1         2                                   GoldenEye (1995)  01-Jan-1995   
2         3                                  Four Rooms (1995)  01-Jan-1995   
3         4                                  Get Shorty (1995)  01-Jan-1995   
4         5                                     Copycat (1995)  01-Jan-1995   
5         6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...  01-Jan-1995   
6         7                              Twelve Monkeys (1995)  01-Jan-1995   
7         8                                        Babe (1995)  01-Jan-1995   
8         9                            Dead Man Walking (1995)  01-Jan-1995   
9        10                                 Richard III (1995)  22-Jan-1996   

   video release date                                           IMDb URL  \
0                 NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                 NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                 NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                 NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                 NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
5                 NaN  http://us.imdb.com/Title?Yao+a+yao+yao+dao+wai...   
6                 NaN  http://us.imdb.com/M/title-exact?Twelve%20Monk...   
7                 NaN     http://us.imdb.com/M/title-exact?Babe%20(1995)   
8                 NaN  http://us.imdb.com/M/title-exact?Dead%20Man%20...   
9                 NaN  http://us.imdb.com/M/title-exact?Richard%20III...   

   unknown  Action  Adventure  Animation  Childrens   ...     Fantasy  \
0        0       0          0          1          1   ...           0   
1        0       1          1          0          0   ...           0   
2        0       0          0          0          0   ...           0   
3        0       1          0          0          0   ...           0   
4        0       0          0          0          0   ...           0   
5        0       0          0          0          0   ...           0   
6        0       0          0          0          0   ...           0   
7        0       0          0          0          1   ...           0   
8        0       0          0          0          0   ...           0   
9        0       0          0          0          0   ...           0   

   Film-Noir  Horror  Musical  Mystery  Romance   Sci-Fi  Thriller  War  \
0          0       0        0        0         0       0         0    0   
1          0       0        0        0         0       0         1    0   
2          0       0        0        0         0       0         1    0   
3          0       0        0        0         0       0         0    0   
4          0       0        0        0         0       0         1    0   
5          0       0        0        0         0       0         0    0   
6          0       0        0        0         0       1         0    0   
7          0       0        0        0         0       0         0    0   
8          0       0        0        0         0       0         0    0   
9          0       0        0        0         0       0         0    1   

   Western  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  

[10 rows x 24 columns]

In [17]:
user[:10]

user id  age gender     occupation zip code
0        1   24      M     technician    85711
1        2   53      F          other    94043
2        3   23      M         writer    32067
3        4   24      M     technician    43537
4        5   33      F          other    15213
5        6   42      M      executive    98101
6        7   57      M  administrator    91344
7        8   36      M  administrator    05201
8        9   29      M        student    01002
9       10   53      M         lawyer    90703

In [22]:
all_data= pd.merge(pd.merge(user, data),item)
all_data[:10]

user id  age gender  occupation zip code  movie id  rating  timestamp  \
0        1   24      M  technician    85711        61       4  878542420   
1       13   47      M    educator    29206        61       4  882140552   
2       18   35      F       other    37212        61       4  880130803   
3       58   27      M  programmer    52246        61       5  884305271   
4       59   49      M    educator    08403        61       4  888204597   
5       60   50      M  healthcare    06472        61       4  883326652   
6       76   20      M     student    02215        61       4  875028123   
7       94   26      M     student    71457        61       5  891720761   
8      144   53      M  programmer    20910        61       3  888106182   
9      154   25      M     student    53703        61       4  879138657   

                  movie title release date   ...     Fantasy Film-Noir  \
0  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
1  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
2  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
3  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
4  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
5  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
6  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
7  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
8  Three Colors: White (1994)  01-Jan-1994   ...           0         0   
9  Three Colors: White (1994)  01-Jan-1994   ...           0         0   

   Horror  Musical  Mystery  Romance   Sci-Fi  Thriller  War  Western  
0       0        0        0         0       0         0    0        0  
1       0        0        0         0       0         0    0        0  
2       0        0        0         0       0         0    0        0  
3       0        0        0         0       0         0    0        0  
4       0        0        0         0       0         0    0        0  
5       0        0        0         0       0         0    0        0  
6       0        0        0         0       0         0    0        0  
7       0        0        0         0       0         0    0        0  
8       0        0        0         0       0         0    0        0  
9       0        0        0         0       0         0    0        0  

[10 rows x 31 columns]

###### check if there is any missing values in the dataset

In [43]:
data.isnull().any()

user id      False
movie id     False
rating       False
timestamp    False
dtype: bool

In [44]:
item.isnull().any()

movie id              False
movie title           False
release date           True
video release date     True
IMDb URL               True
unknown               False
Action                False
Adventure             False
Animation             False
Childrens             False
Comedy                False
Crime                 False
Documentary           False
Drama                 False
Fantasy               False
Film-Noir             False
Horror                False
Musical               False
Mystery               False
Romance               False
Sci-Fi                False
Thriller              False
War                   False
Western               False
dtype: bool

In [45]:
user.isnull().any()

user id       False
age           False
gender        False
occupation    False
zip code      False
dtype: bool

###### check basic summary of the data

###### Summary: There are 943 users and 1682 movies, 10,000 ratings in total

In [50]:
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

In [51]:
data.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

In [52]:
item.describe()

movie id  video release date      unknown       Action    Adventure  \
count  1682.000000                 0.0  1682.000000  1682.000000  1682.000000   
mean    841.500000                 NaN     0.001189     0.149227     0.080262   
std     485.695893                 NaN     0.034473     0.356418     0.271779   
min       1.000000                 NaN     0.000000     0.000000     0.000000   
25%     421.250000                 NaN     0.000000     0.000000     0.000000   
50%     841.500000                 NaN     0.000000     0.000000     0.000000   
75%    1261.750000                 NaN     0.000000     0.000000     0.000000   
max    1682.000000                 NaN     1.000000     1.000000     1.000000   

         Animation    Childrens       Comedy        Crime  Documentary  \
count  1682.000000  1682.000000  1682.000000  1682.000000  1682.000000   
mean      0.024970     0.072533     0.300238     0.064804     0.029727   
std       0.156081     0.259445     0.458498     0.246253     0.169882   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

          ...          Fantasy    Film-Noir       Horror      Musical  \
count     ...       1682.00000  1682.000000  1682.000000  1682.000000   
mean      ...          0.01308     0.014269     0.054697     0.033294   
std       ...          0.11365     0.118632     0.227455     0.179456   
min       ...          0.00000     0.000000     0.000000     0.000000   
25%       ...          0.00000     0.000000     0.000000     0.000000   
50%       ...          0.00000     0.000000     0.000000     0.000000   
75%       ...          0.00000     0.000000     0.000000     0.000000   
max       ...          1.00000     1.000000     1.000000     1.000000   

           Mystery     Romance        Sci-Fi     Thriller          War  \
count  1682.000000  1682.000000  1682.000000  1682.000000  1682.000000   
mean      0.036266     0.146849     0.060048     0.149227     0.042212   
std       0.187008     0.354061     0.237646     0.356418     0.201131   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

           Western  
count  1682.000000  
mean      0.016052  
std       0.125714  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000  

[8 rows x 21 columns]

In [47]:
all_data.describe() # using the merged dataset

user id            age       movie id         rating  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean      462.48475      32.969850     425.530130       3.529860   
std       266.61442      11.562623     330.798356       1.125674   
min         1.00000       7.000000       1.000000       1.000000   
25%       254.00000      24.000000     175.000000       3.000000   
50%       447.00000      30.000000     322.000000       4.000000   
75%       682.00000      40.000000     631.000000       4.000000   
max       943.00000      73.000000    1682.000000       5.000000   

          timestamp  video release date      unknown         Action  \
count  1.000000e+05                 0.0  100000.0000  100000.000000   
mean   8.835289e+08                 NaN       0.0001       0.255890   
std    5.343856e+06                 NaN       0.0100       0.436362   
min    8.747247e+08                 NaN       0.0000       0.000000   
25%    8.794487e+08                 NaN       0.0000       0.000000   
50%    8.828269e+08                 NaN       0.0000       0.000000   
75%    8.882600e+08                 NaN       0.0000       1.000000   
max    8.932866e+08                 NaN       1.0000       1.000000   

           Adventure      Animation      ...              Fantasy  \
count  100000.000000  100000.000000      ...        100000.000000   
mean        0.137530       0.036050      ...             0.013520   
std         0.344408       0.186416      ...             0.115487   
min         0.000000       0.000000      ...             0.000000   
25%         0.000000       0.000000      ...             0.000000   
50%         0.000000       0.000000      ...             0.000000   
75%         0.000000       0.000000      ...             0.000000   
max         1.000000       1.000000      ...             1.000000   

           Film-Noir         Horror        Musical        Mystery  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean        0.017330       0.053170       0.049540       0.052450   
std         0.130498       0.224373       0.216994       0.222934   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

            Romance         Sci-Fi      Thriller            War        Western  
count  100000.000000  100000.00000  100000.00000  100000.000000  100000.000000  
mean        0.194610       0.12730       0.21872       0.093980       0.018540  
std         0.395902       0.33331       0.41338       0.291802       0.134894  
min         0.000000       0.00000       0.00000       0.000000       0.000000  
25%         0.000000       0.00000       0.00000       0.000000       0.000000  
50%         0.000000       0.00000       0.00000       0.000000       0.000000  
75%         0.000000       0.00000       0.00000       0.000000       0.000000  
max         1.000000       1.00000       1.00000       1.000000       1.000000  

[8 rows x 25 columns]

In [41]:
movie_count = all_data.groupby('user id').size() # how many movies each user have rated
movie_count[:10]

user id
1     272
2      62
3      54
4      24
5     175
6     211
7     403
8      59
9      22
10    184
dtype: int64

In [40]:
rating_count = all_data.groupby('movie id').size()# how many ratings each movie have
rating_count[:10]

movie id
1     452
2     131
3      90
4     209
5      86
6      26
7     392
8     219
9     299
10     89
dtype: int64

In [5]:
rating_pivot=data.pivot(index='user id',columns='movie id',values='rating').fillna(0)

In [23]:
rating_pivot[:10]

movie id  1     2     3     4     5     6     7     8     9     10    ...   \
user id                                                               ...    
1          5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...    
2          4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...    
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
5          4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
6          4.0   0.0   0.0   0.0   0.0   0.0   2.0   4.0   4.0   0.0  ...    
7          0.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   4.0  ...    
8          0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0  ...    
9          0.0   0.0   0.0   0.0   0.0   5.0   4.0   0.0   0.0   0.0  ...    
10         4.0   0.0   0.0   4.0   0.0   0.0   4.0   0.0   4.0   0.0  ...    

movie id  1673  1674  1675  1676  1677  1678  1679  1680  1681  1682  
user id                                                               
1          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
7          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
8          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
9          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
10         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[10 rows x 1682 columns]

In [24]:
print rating_pivot.shape

(943, 1682)


In [120]:
rating_pivot.T[:10]

user id   1    2    3    4    5    6    7    8    9    10  ...   934  935  \
movie id                                                   ...              
1         5.0  4.0  0.0  0.0  4.0  4.0  0.0  0.0  0.0  4.0 ...   2.0  3.0   
2         3.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0 ...   4.0  0.0   
3         4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
4         3.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  4.0 ...   5.0  0.0   
5         3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
6         5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0 ...   0.0  0.0   
7         4.0  0.0  0.0  0.0  0.0  2.0  5.0  3.0  4.0  4.0 ...   0.0  0.0   
8         1.0  0.0  0.0  0.0  0.0  4.0  5.0  0.0  0.0  0.0 ...   0.0  0.0   
9         5.0  0.0  0.0  0.0  0.0  4.0  5.0  0.0  0.0  4.0 ...   0.0  1.0   
10        3.0  2.0  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0 ...   0.0  0.0   

user id   936  937  938  939  940  941  942  943  
movie id                                          
1         4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0  
2         0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
3         4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4         0.0  0.0  0.0  0.0  2.0  0.0  0.0  0.0  
5         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
6         5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
7         4.0  0.0  4.0  0.0  4.0  4.0  0.0  0.0  
8         0.0  0.0  0.0  0.0  5.0  0.0  0.0  0.0  
9         4.0  5.0  3.0  5.0  3.0  0.0  0.0  3.0  
10        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[10 rows x 943 columns]

# PART 1: EXPLORATION ON THE DATA with the three algorithms

### collaborative model 1: item based knn implementation

In [7]:
rating_pivot_movie=rating_pivot.T
from scipy.sparse import csr_matrix
rating_matrix_movie=csr_matrix(rating_pivot_movie.values)

In [8]:
from sklearn.neighbors import NearestNeighbors

In [9]:
model_knn=NearestNeighbors(metric='cosine',algorithm='brute')
model_knn.fit(rating_matrix_movie)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
         metric_params=None, n_jobs=1, n_neighbors=5, p=2, radius=1.0)

In [10]:
query_index=np.random.choice(rating_pivot_movie.shape[0])

In [11]:
print query_index

1547


###### look for movie title in the item dataset and return in the query result below:

In [12]:
distances, indices = model_knn.kneighbors(rating_pivot_movie.iloc[query_index, :].reshape(1, -1), n_neighbors = 6)
# look for movie title in the item dataset and return in the query result below:
for i in range(0, len(distances.flatten())):
    if i == 0:
        print 'Recommendations for {0}:\n'.format(item.loc[item['movie id'] ==rating_pivot_movie.index[query_index],'movie title'].item() )
    else:
        print '{0}: {1}, with distance of {2}:'.format(i, item.loc[item['movie id'] ==rating_pivot_movie.index[indices.flatten()[i]],'movie title'].item(), distances.flatten()[i])

Recommendations for The Courtyard (1995):

1: Quartier Mozart (1992), with distance of 0.0:
2: I, Worst of All (Yo, la peor de todas) (1990), with distance of 0.0:
3: Touki Bouki (Journey of the Hyena) (1973), with distance of 0.0:
4: The Courtyard (1995), with distance of 0.0:
5: Pharaoh's Army (1995), with distance of 0.0:


/anaconda/lib/python2.7/site-packages/ipykernel_launcher.py:1: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  """Entry point for launching an IPython kernel.


In [143]:
distances # we are looking at 6 neighbors, and the first one is the distance to itself, that is why it is 0.

array([[ 0.        ,  0.5119975 ,  0.51421624,  0.52473047,  0.5777278 ,
         0.58980777]])

### collaborative model 2: user based knn implementation

In [14]:
vote=np.array(rating_pivot)
vote_mean=vote.mean(axis=1)

In [15]:
from sklearn.neighbors import NearestNeighbors
model_knn2=NearestNeighbors(metric='cosine',algorithm='brute')
model_knn2.fit(vote)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
         metric_params=None, n_jobs=1, n_neighbors=5, p=2, radius=1.0)

In [16]:
query_index2=np.random.choice(vote.shape[0])

In [17]:
print query_index2

902


In [18]:
distances2, indices2 = model_knn2.kneighbors(vote[query_index2].reshape(1, -1), n_neighbors = 6)
# look for movie title in the item dataset and return in the query result below:
for i in range(0, len(distances2.flatten())):
    if i == 0:
        print 'Recommendations for user id {0}:\n'.format(query_index2)
    else:
        print '{0}: user id {1}, with distance of {2}:'.format(i, indices2.flatten()[i], distances2.flatten()[i])

Recommendations for user id 902:

1: user id 193, with distance of 0.550999338662:
2: user id 176, with distance of 0.55512852004:
3: user id 473, with distance of 0.557581364879:
4: user id 915, with distance of 0.557891291468:
5: user id 665, with distance of 0.560886530711:


In [19]:
indices2

array([[902, 193, 176, 473, 915, 665]])

In [20]:
distances2

array([[ 0.        ,  0.55099934,  0.55512852,  0.55758136,  0.55789129,
         0.56088653]])

###### Given a record user id 298, moive id 474, what would be the user's rating for this movie?

###### user_based method would be: calculate nearest k users using pivot with all users who have rated this movie and without column 474, as we assume this info is unknow for user 298;  as  column movie 474 would be saved as target and average the nearest k neighbor's rating on this movie

In [ ]:
# movie_566=data.loc[(data["movie id"] == 566) & (d["C"] == 900), "user id"]

In [133]:
i=474
rating_i=rating_pivot.loc[rating_pivot[i] != 0]
print rating_i.shape

(194, 1682)


In [144]:
rating_i.index.values

array([  6,   7,  10,  13,  14,  18,  25,  41,  44,  58,  59,  60,  62,
        73,  76,  77,  82,  85,  90,  91,  92,  94,  95,  96, 114, 118,
       124, 125, 138, 144, 148, 151, 154, 160, 162, 188, 189, 194, 198,
       213, 215, 226, 232, 234, 235, 237, 239, 248, 250, 267, 268, 269,
       271, 272, 276, 279, 290, 293, 297, 298, 299, 301, 303, 305, 307,
       312, 313, 318, 321, 325, 326, 327, 328, 334, 338, 339, 343, 360,
       363, 373, 379, 380, 382, 383, 385, 387, 389, 391, 397, 398, 406,
       407, 409, 417, 421, 425, 426, 433, 435, 447, 450, 452, 454, 456,
       457, 458, 465, 469, 474, 479, 487, 488, 492, 498, 499, 508, 514,
       516, 521, 524, 527, 533, 536, 537, 541, 543, 545, 553, 561, 567,
       568, 603, 607, 622, 640, 643, 645, 648, 650, 653, 655, 659, 660,
       666, 669, 670, 686, 694, 716, 727, 737, 738, 747, 748, 757, 758,
       766, 776, 780, 781, 788, 796, 804, 823, 830, 833, 840, 843, 846,
       847, 848, 862, 864, 867, 868, 870, 875, 878, 881, 886, 88

In [145]:
j=298
record=rating_i.loc[rating_i.index.values==j].values[0]
print record

[ 5.  0.  0. ...,  0.  0.  0.]


In [146]:
target=rating_i[i]
print target.shape
temp=rating_i.drop(i, 1)
print temp.shape
train=np.array(temp)
train_mean=train.mean(axis=1)

(194,)
(194, 1681)


In [149]:
j=298
record=temp.loc[temp.index.values==j].values[0]
print record

[ 5.  0.  0. ...,  0.  0.  0.]


In [150]:
from sklearn.neighbors import NearestNeighbors
model_knn3=NearestNeighbors(metric='cosine',algorithm='brute')
model_knn3.fit(train)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
         metric_params=None, n_jobs=1, n_neighbors=5, p=2, radius=1.0)

In [166]:
distances3, indices3 = model_knn3.kneighbors(record.reshape(1, -1), n_neighbors =6)

In [182]:
rate=0
counter=0
unique_index=rating_i.index.values
user_ids=[]
n_neighbors=6
original_rating= data.ix[(data['user id'] ==j) & (data['movie id']==i)].rating.item()
# look for movie title in the item dataset and return in the query result below:
for k in range(0, len(distances3.flatten())):
    user_id=unique_index[indices3[0][k]]
    user_ids.append(user_id)
    if i == 0:
        print 'Recommendations for user id {0}:\n'.format(j)
    else:
        newdf=data.ix[(data['user id'] ==user_id) & (data['movie id']==i)]
        if newdf.empty:
            pass
        else:
            rating=newdf.rating.item()
            counter=counter+1
            rate=rate+rating
average_rate=rate/counter
print 'Predicted rating of movie {0} for user id {1} is: {2}'.format(i,j,average_rate)
print 'The original rating of movie {0} for user id {1} is: {2} '.format(i,j,original_rating)
print "The nearest {0} neighbors are(including itself):".format(n_neighbors)
print user_ids

Predicted rating of movie 474 for user id 298 is: 4
The original rating of movie 474 for user id 298 is: 4 
The nearest 6 neighbors are(including itself):
[298, 716, 313, 194, 533, 271]


### collaborative model 3: matrix factorization implementation

###### We have tried to predict the rating by user and by movie in the last two models, now we could try to recommend new movies to user to watch using matrix factorization

In [183]:
#normalize the data by subtracting each user rating by the each user mean
temp=rating_pivot.as_matrix()
user_rating_mean=np.mean(temp,axis=1)
data_demeaned=temp-user_rating_mean.reshape(-1,1)
data_demeaned[:2]

array([[ 4.41617122,  2.41617122,  3.41617122, ..., -0.58382878,
        -0.58382878, -0.58382878],
       [ 3.86325803, -0.13674197, -0.13674197, ..., -0.13674197,
        -0.13674197, -0.13674197]])

In [184]:
from scipy.sparse.linalg import svds
U, sigma, Vt = svds(data_demeaned, k = 50)
sigma=np.diag(sigma)

In [185]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) + user_rating_mean.reshape(-1, 1)
preds_df = pd.DataFrame(all_user_predicted_ratings, columns = rating_pivot.columns)

In [186]:
def recommend_movies(predictions_df, userID, movies_df, original_ratings_df, num_recommendations=5):
    
    # Get and sort the user's predictions
    user_row_number = userID - 1 # UserID starts at 1, not 0
    sorted_user_predictions = predictions_df.iloc[user_row_number].sort_values(ascending=False)
    
    # Get the user's data and merge in the movie information.
    user_data = original_ratings_df[original_ratings_df['user id'] == (userID)]
    user_full = (user_data.merge(movies_df, how = 'left', left_on = 'movie id', right_on = 'movie id').
                     sort_values(['rating'], ascending=False)
                 )

    print 'User {0} has already rated {1} movies.'.format(userID, user_full.shape[0])
    print 'Recommending the highest {0} predicted ratings movies not already rated.'.format(num_recommendations)
    
    # Recommend the highest predicted rating movies that the user hasn't seen yet.
    recommendations = (movies_df[~movies_df['movie id'].isin(user_full['movie id'])].
         merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
               left_on = 'movie id',
               right_on = 'movie id').
         rename(columns = {user_row_number: 'Predictions'}).
         sort_values('Predictions', ascending = False).
                       iloc[:num_recommendations, :-1]
                      )

    return user_full, recommendations

In [191]:
already_rated, predictions = recommend_movies(preds_df, 49, item, data, 5)

User 49 has already rated 215 movies.
Recommending the highest 5 predicted ratings movies not already rated.


In [192]:
already_rated.head(10)

user id  movie id  rating  timestamp  \
132       49       433       5  888068739   
106       49       179       5  888066446   
55        49       151       5  888067727   
66        49       657       5  888068032   
38        49        56       5  888067307   
37        49       168       5  888068686   
176       49       209       5  888068877   
175       49       919       5  888066133   
168       49        93       5  888068912   
167       49       432       5  888066979   

                                      movie title release date  \
132                               Heathers (1989)  01-Jan-1989   
106                    Clockwork Orange, A (1971)  01-Jan-1971   
55   Willy Wonka and the Chocolate Factory (1971)  01-Jan-1971   
66               Manchurian Candidate, The (1962)  01-Jan-1962   
38                            Pulp Fiction (1994)  01-Jan-1994   
37         Monty Python and the Holy Grail (1974)  01-Jan-1974   
176                     This Is Spinal Tap (1984)  01-Jan-1984   
175             City of Lost Children, The (1995)  01-Jan-1995   
168               Welcome to the Dollhouse (1995)  24-May-1996   
167                               Fantasia (1940)  01-Jan-1940   

     video release date                                           IMDb URL  \
132                 NaN  http://us.imdb.com/M/title-exact?Heathers%20(1...   
106                 NaN  http://us.imdb.com/M/title-exact?Clockwork%20O...   
55                  NaN  http://us.imdb.com/M/title-exact?Willy%20Wonka...   
66                  NaN  http://us.imdb.com/M/title-exact?Manchurian%20...   
38                  NaN  http://us.imdb.com/M/title-exact?Pulp%20Fictio...   
37                  NaN  http://us.imdb.com/M/title-exact?Monty%20Pytho...   
176                 NaN  http://us.imdb.com/M/title-exact?This%20Is%20S...   
175                 NaN  http://us.imdb.com/Title?Cit%E9+des+enfants+pe...   
168                 NaN  http://us.imdb.com/Title?Welcome+to+the+Dollho...   
167                 NaN  http://us.imdb.com/M/title-exact?Fantasia%20(1...   

     unknown  Action   ...     Fantasy  Film-Noir  Horror  Musical  Mystery  \
132        0       0   ...           0          0       0        0        0   
106        0       0   ...           0          0       0        0        0   
55         0       0   ...           0          0       0        0        0   
66         0       0   ...           0          1       0        0        0   
38         0       0   ...           0          0       0        0        0   
37         0       0   ...           0          0       0        0        0   
176        0       0   ...           0          0       0        1        0   
175        0       0   ...           0          0       0        0        0   
168        0       0   ...           0          0       0        0        0   
167        0       0   ...           0          0       0        1        0   

     Romance   Sci-Fi  Thriller  War  Western  
132         0       0         0    0        0  
106         0       1         0    0        0  
55          0       0         0    0        0  
66          0       0         1    0        0  
38          0       0         0    0        0  
37          0       0         0    0        0  
176         0       0         0    0        0  
175         0       1         0    0        0  
168         0       0         0    0        0  
167         0       0         0    0        0  

[10 rows x 27 columns]

In [193]:
predictions

movie id                                        movie title release date  \
568       742                                      Ransom (1996)  08-Nov-1996   
97        156                              Reservoir Dogs (1992)  01-Jan-1992   
412       558                          Heavenly Creatures (1994)  01-Jan-1994   
295       421      William Shakespeare's Romeo and Juliet (1996)  25-Oct-1996   
340       474  Dr. Strangelove or: How I Learned to Stop Worr...  01-Jan-1963   

     video release date                                           IMDb URL  \
568                 NaN   http://us.imdb.com/M/title-exact?Ransom%20(1996)   
97                  NaN  http://us.imdb.com/M/title-exact?Reservoir%20D...   
412                 NaN  http://us.imdb.com/M/title-exact?Heavenly%20Cr...   
295                 NaN   http://us.imdb.com/Title?Romeo+%2B+Juliet+(1996)   
340                 NaN  http://us.imdb.com/M/title-exact?Dr.%20Strange...   

     unknown  Action  Adventure  Animation  Childrens   ...     Fantasy  \
568        0       0          0          0          0   ...           0   
97         0       0          0          0          0   ...           0   
412        0       0          0          0          0   ...           1   
295        0       0          0          0          0   ...           0   
340        0       0          0          0          0   ...           0   

     Film-Noir  Horror  Musical  Mystery  Romance   Sci-Fi  Thriller  War  \
568          0       0        0        0         0       0         1    0   
97           0       0        0        0         0       0         1    0   
412          0       0        0        0         0       0         1    0   
295          0       0        0        0         1       0         0    0   
340          0       0        0        0         0       1         0    1   

     Western  
568        0  
97         0  
412        0  
295        0  
340        0  

[5 rows x 24 columns]

# PART 2: Rating accuracy Comparsion for the 3 algorithms

###### now we use the test data from the dataset, ua.test, the data will be used to bulid models and the test data will be used for accuracy and efficiency comparison among the 3 different models: user-based knn, item based knn, and matrix factorization

In [357]:
test=pd.read_csv('ml-100k/ua.test', sep='\t', names=dataCol, encoding='latin-1')

In [358]:
test.shape

(9430, 4)

In [487]:
test.head

<bound method DataFrame.head of       user id  movie id  rating  timestamp
0           1        20       4  887431883
1           1        33       4  878542699
2           1        61       4  878542420
3           1       117       3  874965739
4           1       155       2  878542201
5           1       160       4  875072547
6           1       171       5  889751711
7           1       189       3  888732928
8           1       202       5  875072442
9           1       265       4  878542441
10          2        13       4  888551922
11          2        50       5  888552084
12          2       251       5  888552084
13          2       280       3  888551441
14          2       281       3  888980240
15          2       290       3  888551441
16          2       292       4  888550774
17          2       297       4  888550871
18          2       312       3  888550631
19          2       314       1  888980085
20          3       245       1  889237247
21          3       29

## collaborative model 1: item based knn implementation

###### for each test instance, in the movies that the user have rated, find the movies closest with the instance movie

In [391]:
rating_pivot_movie=rating_pivot.T
#rating_i=rating_pivot.loc[rating_pivot[i] != 0]
from scipy.sparse import csr_matrix
rating_matrix_movie=csr_matrix(rating_pivot_movie.values)

In [393]:
rating_pivot_movie[:10]
print rating_pivot_movie.shape


(1682, 943)


In [397]:
rating_pivot_movie[:5]

user id   1    2    3    4    5    6    7    8    9    10  ...   934  935  \
movie id                                                   ...              
1         5.0  4.0  0.0  0.0  4.0  4.0  0.0  0.0  0.0  4.0 ...   2.0  3.0   
2         3.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0 ...   4.0  0.0   
3         4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
4         3.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  4.0 ...   5.0  0.0   
5         3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   

user id   936  937  938  939  940  941  942  943  
movie id                                          
1         4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0  
2         0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
3         4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4         0.0  0.0  0.0  0.0  2.0  0.0  0.0  0.0  
5         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 943 columns]

In [416]:
#rating_i=rating_pivot_movie.loc[rating_pivot_movie[936] != 0]
#rating_i.describe()
#rating_i[:5]
#rating_i.shape

(142, 943)

In [394]:
from sklearn.neighbors import NearestNeighbors

In [395]:
model_knn=NearestNeighbors(metric='cosine',algorithm='brute')
model_knn.fit(rating_matrix_movie)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
         metric_params=None, n_jobs=1, n_neighbors=5, p=2, radius=1.0)

In [497]:
array_test=test.values
array_test[:,2]
classes=array_test[:,2].tolist()

In [579]:
from time import time
t0=time()
rawRating=[] # for raw averaged rating from k nearest neighbors
ratingFortest=[] # fro rounded average rating from k nearest neighbors
for id in array_test:
    #print id
    query_index=id[1] #movie
    user_id=id[0] # user
    neighbors=[]
    rating=0.0
    counter=0
    rating_i=rating_pivot_movie.loc[rating_pivot_movie[user_id] != 0]
    rating_matrix_movie=csr_matrix(rating_i.values)
    #print rating_i[:5]
    model_knn=NearestNeighbors(metric='cosine',algorithm='brute')
    model_knn.fit(rating_matrix_movie)
    unique_index=rating_i.index.values
    #print unique_index
    if query_index in unique_index:
        distances, indices = model_knn.kneighbors(rating_i.loc[[query_index]].values, n_neighbors = 6)
        #print indices
        for i in range(0, len(indices.flatten())):
            if i == 0:
                pass
            else:
                movie_id=unique_index[indices[0][i]]
                rate=data.loc[((data['user id'] == user_id) & (data['movie id'] == movie_id))].rating.item()
                rating=rating+rate
                counter=counter+1    
        avg_rating=rating/counter        
    else:
        avg_rating=rating_i[user_id].mean()
    rawRating.append(avg_rating)
    ratingFortest.append(round(avg_rating))
print('done in %0.3fs.'%(time()-t0))

done in 102.042s.


In [580]:
ratingFortest[:10]

[5.0, 3.0, 5.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0]

In [450]:
from sklearn.metrics import completeness_score, homogeneity_score

In [464]:
print completeness_score(array_test[:,2],ratingFortest)

0.104681965465


In [465]:
print homogeneity_score(array_test[:,2],ratingFortest)

0.0790519777136


In [582]:
err=0.0
for i in range(9430):
    error=abs(classes[i]-rawRating[i])/classes[i]
    err=err+error
print 'The overall mean absolute error(MAE) of prediction is: '
print err/9430

The overall mean absolute error(MAE) of prediction is: 
0.314492753623


## Collaborative model 2: user based knn implementation

In [ ]:
# find the most similar k users who have rated this movie, and get a average

In [ ]:
# for each record in test, find the subset of ratings of all the ratings for this movie

In [472]:
rating_pivot[:10]

movie id  1     2     3     4     5     6     7     8     9     10    ...   \
user id                                                               ...    
1          5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...    
2          4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...    
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
5          4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
6          4.0   0.0   0.0   0.0   0.0   0.0   2.0   4.0   4.0   0.0  ...    
7          0.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   4.0  ...    
8          0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0  ...    
9          0.0   0.0   0.0   0.0   0.0   5.0   4.0   0.0   0.0   0.0  ...    
10         4.0   0.0   0.0   4.0   0.0   0.0   4.0   0.0   4.0   0.0  ...    

movie id  1673  1674  1675  1676  1677  1678  1679  1680  1681  1682  
user id                                                               
1          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
7          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
8          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
9          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
10         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[10 rows x 1682 columns]

In [477]:
rating_pivot.shape[0]

943

In [584]:
from time import time
t0=time()
rawRating=[]
ratingFortest=[]
for id in array_test:
    #print id
    query_index=id[1] #movie
    user_id=id[0] # user
    neighbors=[]
    rating=0.0
    counter=0
    rating_i=rating_pivot.loc[rating_pivot[query_index] != 0]
    rating_matrix_movie=csr_matrix(rating_i.values)
    #print rating_i[:5]
    model_knn=NearestNeighbors(metric='cosine',algorithm='brute')
    model_knn.fit(rating_matrix_movie)
    unique_index=rating_i.index.values
    #print unique_index
    if rating_i.shape[0]<4:
        avg_rating=rating_i[query_index].mean()
    else:     
        distances, indices = model_knn.kneighbors(rating_i.loc[[user_id]].values, n_neighbors = 4)
        #print indices
        for i in range(0, len(indices.flatten())):
            if i == 0:
                pass
            else:
                k_user_id=unique_index[indices[0][i]]
                rate=data.loc[((data['user id'] == k_user_id) & (data['movie id'] == query_index))].rating.item()
                rating=rating+rate
                counter=counter+1    
        avg_rating=rating/counter        
    rawRating.append(avg_rating)
    ratingFortest.append(round(avg_rating))
print('done in %0.3fs.'%(time()-t0))

done in 159.563s.


In [508]:
ratingFortest[:20]
labels=[int(i) for i in ratingFortest]
labels[:20]

[4, 3, 4, 3, 2, 4, 4, 4, 4, 4, 3, 4, 4, 3, 3, 3, 4, 4, 4, 1]

In [511]:
classes[:20]

[4, 4, 4, 3, 2, 4, 5, 3, 5, 4, 4, 5, 5, 3, 3, 3, 4, 4, 3, 1]

In [507]:
print completeness_score(classes,labels)

0.0547050495129


In [499]:
print homogeneity_score(classes,ratingFortest)

0.0450807342562


In [531]:
err=0.0
for i in range(9430):
    error=abs(classes[i]-rawRating[i])
    err=err+error
print 'The overall mean absolute error(MAE) of prediction is: '
print err/9430

The overall mean absolute error(MAE) of prediction is: 
0.330851891128
The overall accracy of prediction is:
0.669148108872


0

## Collaborative model 3: matrix factorization implementation

###### FIRSTLY, import some predefined functions of matrix factorizaiton from assignment 4

In [545]:
def matrix_factorization(R,P,Q,K,steps=5000,alpha=0.0002,beta=0.02):
    Q=Q.T
    for step in xrange(steps):
        for i in xrange(len(R)):
            for j in xrange(len(R[i])):
                if R[i][j]>0:
                    eij=R[i][j]-np.dot(P[i,:],Q[:,j])
                    for k in xrange(K):
                        P[i][k]=P[i][k]+alpha*(2*eij*Q[k][j]-beta*P[i][k])
                        Q[k][j]=Q[k][j]+alpha*(2*eij*P[i][k]-beta*Q[k][j])
        eR=np.dot(P,Q)
        e=0
        for i in xrange(len(R)):
            for j in xrange(len(R[i])):
                if R[i][j]>0:
                    e=e+pow(R[i][j]-np.dot(P[i,:],Q[:,j]),2)
                    for k in xrange(K):
                        e=e+(beta/2)*(pow(P[i][k],2)+pow(Q[k][j],2))
        if e<0.001:
            break
        print "step %d of %d; error: %0.5f; time:%0.2f"%(step+1,steps,e,time())
    return P,Q.T

In [539]:
dataMat=rating_pivot.values
M=len(data)
N=len(data[0])
print M, N
K=5
steps=3000

943 1682


In [541]:
P=np.random.rand(M,K)
Q=np.random.rand(N,K)

In [546]:
from time import time
t0=time()
fP,fQ=matrix_factorization(dataMat,P,Q,K,steps=steps)
print('done in %0.3fs.'%(time()-t0))

step 1 of 3000; error: 355215.40269; time:1510535739.42
step 2 of 3000; error: 268500.15583; time:1510535744.16
step 3 of 3000; error: 217480.09913; time:1510535748.64
step 4 of 3000; error: 187321.52107; time:1510535753.01
step 5 of 3000; error: 168261.05227; time:1510535757.36
step 6 of 3000; error: 155197.14249; time:1510535761.96
step 7 of 3000; error: 145632.61028; time:1510535766.46
step 8 of 3000; error: 138297.25278; time:1510535770.79
step 9 of 3000; error: 132486.44198; time:1510535775.34
step 10 of 3000; error: 127772.70212; time:1510535779.80
step 11 of 3000; error: 123877.62125; time:1510535784.28
step 12 of 3000; error: 120610.33785; time:1510535788.75
step 13 of 3000; error: 117835.04247; time:1510535793.01
step 14 of 3000; error: 115452.29725; time:1510535797.73
step 15 of 3000; error: 113387.58696; time:1510535802.81
step 16 of 3000; error: 111583.96685; time:1510535808.66
step 17 of 3000; error: 109997.17377; time:1510535813.85
step 18 of 3000; error: 108592.28439; ti

step 147 of 3000; error: 89632.77558; time:1510536398.21
step 148 of 3000; error: 89603.16093; time:1510536402.81
step 149 of 3000; error: 89573.52401; time:1510536407.33
step 150 of 3000; error: 89543.86007; time:1510536411.76
step 151 of 3000; error: 89514.16447; time:1510536416.22
step 152 of 3000; error: 89484.43266; time:1510536420.72
step 153 of 3000; error: 89454.66022; time:1510536425.31
step 154 of 3000; error: 89424.84277; time:1510536429.88
step 155 of 3000; error: 89394.97609; time:1510536434.57
step 156 of 3000; error: 89365.05601; time:1510536439.02
step 157 of 3000; error: 89335.07847; time:1510536443.47
step 158 of 3000; error: 89305.03949; time:1510536447.92
step 159 of 3000; error: 89274.93520; time:1510536452.52
step 160 of 3000; error: 89244.76179; time:1510536457.48
step 161 of 3000; error: 89214.51555; time:1510536462.35
step 162 of 3000; error: 89184.19288; time:1510536467.21
step 163 of 3000; error: 89153.79022; time:1510536472.09
step 164 of 3000; error: 89123.

step 291 of 3000; error: 84389.01171; time:1510537054.55
step 292 of 3000; error: 84350.19079; time:1510537059.03
step 293 of 3000; error: 84311.46676; time:1510537064.08
step 294 of 3000; error: 84272.84141; time:1510537068.87
step 295 of 3000; error: 84234.31648; time:1510537073.16
step 296 of 3000; error: 84195.89360; time:1510537078.10
step 297 of 3000; error: 84157.57438; time:1510537082.79
step 298 of 3000; error: 84119.36034; time:1510537087.40
step 299 of 3000; error: 84081.25295; time:1510537092.29
step 300 of 3000; error: 84043.25361; time:1510537097.11
step 301 of 3000; error: 84005.36366; time:1510537101.72
step 302 of 3000; error: 83967.58438; time:1510537106.48
step 303 of 3000; error: 83929.91700; time:1510537111.45
step 304 of 3000; error: 83892.36267; time:1510537116.40
step 305 of 3000; error: 83854.92249; time:1510537121.17
step 306 of 3000; error: 83817.59752; time:1510537126.23
step 307 of 3000; error: 83780.38874; time:1510537131.07
step 308 of 3000; error: 83743.

step 435 of 3000; error: 79992.23143; time:1510537730.89
step 436 of 3000; error: 79969.31358; time:1510537735.15
step 437 of 3000; error: 79946.48120; time:1510537739.35
step 438 of 3000; error: 79923.73391; time:1510537743.57
step 439 of 3000; error: 79901.07132; time:1510537747.76
step 440 of 3000; error: 79878.49307; time:1510537752.00
step 441 of 3000; error: 79855.99877; time:1510537756.20
step 442 of 3000; error: 79833.58805; time:1510537760.54
step 443 of 3000; error: 79811.26054; time:1510537765.90
step 444 of 3000; error: 79789.01586; time:1510537770.77
step 445 of 3000; error: 79766.85366; time:1510537775.28
step 446 of 3000; error: 79744.77356; time:1510537779.77
step 447 of 3000; error: 79722.77519; time:1510537785.37
step 448 of 3000; error: 79700.85820; time:1510537790.03
step 449 of 3000; error: 79679.02223; time:1510537794.26
step 450 of 3000; error: 79657.26691; time:1510537798.48
step 451 of 3000; error: 79635.59188; time:1510537802.74
step 452 of 3000; error: 79613.

step 579 of 3000; error: 77414.00696; time:1510538364.97
step 580 of 3000; error: 77400.23156; time:1510538369.71
step 581 of 3000; error: 77386.50189; time:1510538374.48
step 582 of 3000; error: 77372.81776; time:1510538378.98
step 583 of 3000; error: 77359.17899; time:1510538383.52
step 584 of 3000; error: 77345.58538; time:1510538387.87
step 585 of 3000; error: 77332.03674; time:1510538392.87
step 586 of 3000; error: 77318.53289; time:1510538397.76
step 587 of 3000; error: 77305.07363; time:1510538402.17
step 588 of 3000; error: 77291.65878; time:1510538406.73
step 589 of 3000; error: 77278.28816; time:1510538411.15
step 590 of 3000; error: 77264.96158; time:1510538415.60
step 591 of 3000; error: 77251.67886; time:1510538420.29
step 592 of 3000; error: 77238.43981; time:1510538424.72
step 593 of 3000; error: 77225.24425; time:1510538429.16
step 594 of 3000; error: 77212.09200; time:1510538433.45
step 595 of 3000; error: 77198.98288; time:1510538438.03
step 596 of 3000; error: 77185.

step 723 of 3000; error: 75821.34281; time:1510539115.12
step 724 of 3000; error: 75812.55076; time:1510539120.01
step 725 of 3000; error: 75803.78425; time:1510539125.23
step 726 of 3000; error: 75795.04319; time:1510539130.19
step 727 of 3000; error: 75786.32747; time:1510539135.12
step 728 of 3000; error: 75777.63700; time:1510539140.23
step 729 of 3000; error: 75768.97167; time:1510539145.15
step 730 of 3000; error: 75760.33139; time:1510539149.85
step 731 of 3000; error: 75751.71607; time:1510539154.33
step 732 of 3000; error: 75743.12560; time:1510539158.70
step 733 of 3000; error: 75734.55989; time:1510539163.21
step 734 of 3000; error: 75726.01884; time:1510539167.66
step 735 of 3000; error: 75717.50236; time:1510539172.24
step 736 of 3000; error: 75709.01035; time:1510539176.47
step 737 of 3000; error: 75700.54272; time:1510539180.65
step 738 of 3000; error: 75692.09938; time:1510539184.83
step 739 of 3000; error: 75683.68022; time:1510539189.11
step 740 of 3000; error: 75675.

step 867 of 3000; error: 74776.34833; time:1510539804.20
step 868 of 3000; error: 74770.39028; time:1510539809.51
step 869 of 3000; error: 74764.44706; time:1510539814.71
step 870 of 3000; error: 74758.51862; time:1510539819.88
step 871 of 3000; error: 74752.60490; time:1510539825.17
step 872 of 3000; error: 74746.70585; time:1510539830.43
step 873 of 3000; error: 74740.82141; time:1510539835.53
step 874 of 3000; error: 74734.95154; time:1510539840.54
step 875 of 3000; error: 74729.09618; time:1510539845.84
step 876 of 3000; error: 74723.25527; time:1510539851.09
step 877 of 3000; error: 74717.42877; time:1510539856.55
step 878 of 3000; error: 74711.61662; time:1510539861.74
step 879 of 3000; error: 74705.81877; time:1510539866.69
step 880 of 3000; error: 74700.03518; time:1510539871.63
step 881 of 3000; error: 74694.26578; time:1510539876.89
step 882 of 3000; error: 74688.51052; time:1510539881.98
step 883 of 3000; error: 74682.76937; time:1510539887.33
step 884 of 3000; error: 74677.

step 1011 of 3000; error: 74048.29606; time:1510540460.50
step 1012 of 3000; error: 74044.01464; time:1510540464.69
step 1013 of 3000; error: 74039.74223; time:1510540468.87
step 1014 of 3000; error: 74035.47879; time:1510540473.04
step 1015 of 3000; error: 74031.22429; time:1510540477.21
step 1016 of 3000; error: 74026.97871; time:1510540481.43
step 1017 of 3000; error: 74022.74202; time:1510540485.61
step 1018 of 3000; error: 74018.51418; time:1510540489.77
step 1019 of 3000; error: 74014.29518; time:1510540493.92
step 1020 of 3000; error: 74010.08497; time:1510540498.09
step 1021 of 3000; error: 74005.88353; time:1510540502.25
step 1022 of 3000; error: 74001.69084; time:1510540506.42
step 1023 of 3000; error: 73997.50686; time:1510540510.57
step 1024 of 3000; error: 73993.33157; time:1510540514.77
step 1025 of 3000; error: 73989.16494; time:1510540518.92
step 1026 of 3000; error: 73985.00694; time:1510540523.11
step 1027 of 3000; error: 73980.85755; time:1510540527.29
step 1028 of 3

step 1153 of 3000; error: 73518.52816; time:1510541054.19
step 1154 of 3000; error: 73515.28094; time:1510541058.41
step 1155 of 3000; error: 73512.03957; time:1510541062.59
step 1156 of 3000; error: 73508.80403; time:1510541066.76
step 1157 of 3000; error: 73505.57430; time:1510541070.95
step 1158 of 3000; error: 73502.35037; time:1510541075.12
step 1159 of 3000; error: 73499.13222; time:1510541079.31
step 1160 of 3000; error: 73495.91984; time:1510541083.48
step 1161 of 3000; error: 73492.71320; time:1510541087.65
step 1162 of 3000; error: 73489.51230; time:1510541091.81
step 1163 of 3000; error: 73486.31711; time:1510541096.03
step 1164 of 3000; error: 73483.12763; time:1510541100.21
step 1165 of 3000; error: 73479.94383; time:1510541104.38
step 1166 of 3000; error: 73476.76570; time:1510541108.56
step 1167 of 3000; error: 73473.59322; time:1510541112.81
step 1168 of 3000; error: 73470.42638; time:1510541116.99
step 1169 of 3000; error: 73467.26517; time:1510541121.15
step 1170 of 3

step 1295 of 3000; error: 73109.22127; time:1510541649.33
step 1296 of 3000; error: 73106.66604; time:1510541653.52
step 1297 of 3000; error: 73104.11485; time:1510541657.77
step 1298 of 3000; error: 73101.56770; time:1510541661.96
step 1299 of 3000; error: 73099.02458; time:1510541666.13
step 1300 of 3000; error: 73096.48548; time:1510541670.29
step 1301 of 3000; error: 73093.95038; time:1510541674.46
step 1302 of 3000; error: 73091.41929; time:1510541678.65
step 1303 of 3000; error: 73088.89218; time:1510541682.84
step 1304 of 3000; error: 73086.36905; time:1510541687.01
step 1305 of 3000; error: 73083.84989; time:1510541691.17
step 1306 of 3000; error: 73081.33469; time:1510541695.36
step 1307 of 3000; error: 73078.82344; time:1510541699.53
step 1308 of 3000; error: 73076.31613; time:1510541703.69
step 1309 of 3000; error: 73073.81275; time:1510541707.89
step 1310 of 3000; error: 73071.31330; time:1510541712.06
step 1311 of 3000; error: 73068.81776; time:1510541716.24
step 1312 of 3

step 1437 of 3000; error: 72782.85096; time:1510542245.53
step 1438 of 3000; error: 72780.78696; time:1510542249.69
step 1439 of 3000; error: 72778.72591; time:1510542253.91
step 1440 of 3000; error: 72776.66780; time:1510542258.14
step 1441 of 3000; error: 72774.61263; time:1510542262.31
step 1442 of 3000; error: 72772.56039; time:1510542266.47
step 1443 of 3000; error: 72770.51106; time:1510542270.63
step 1444 of 3000; error: 72768.46466; time:1510542274.79
step 1445 of 3000; error: 72766.42116; time:1510542278.95
step 1446 of 3000; error: 72764.38057; time:1510542283.13
step 1447 of 3000; error: 72762.34288; time:1510542287.30
step 1448 of 3000; error: 72760.30808; time:1510542291.47
step 1449 of 3000; error: 72758.27617; time:1510542295.67
step 1450 of 3000; error: 72756.24714; time:1510542299.87
step 1451 of 3000; error: 72754.22098; time:1510542304.02
step 1452 of 3000; error: 72752.19769; time:1510542308.18
step 1453 of 3000; error: 72750.17726; time:1510542312.38
step 1454 of 3

step 1579 of 3000; error: 72516.64650; time:1510542841.76
step 1580 of 3000; error: 72514.94654; time:1510542846.01
step 1581 of 3000; error: 72513.24879; time:1510542850.21
step 1582 of 3000; error: 72511.55327; time:1510542854.50
step 1583 of 3000; error: 72509.85995; time:1510542858.76
step 1584 of 3000; error: 72508.16885; time:1510542862.99
step 1585 of 3000; error: 72506.47994; time:1510542867.18
step 1586 of 3000; error: 72504.79324; time:1510542871.40
step 1587 of 3000; error: 72503.10873; time:1510542875.60
step 1588 of 3000; error: 72501.42642; time:1510542879.81
step 1589 of 3000; error: 72499.74629; time:1510542884.06
step 1590 of 3000; error: 72498.06834; time:1510542888.27
step 1591 of 3000; error: 72496.39257; time:1510542892.47
step 1592 of 3000; error: 72494.71898; time:1510542896.74
step 1593 of 3000; error: 72493.04756; time:1510542900.94
step 1594 of 3000; error: 72491.37830; time:1510542905.17
step 1595 of 3000; error: 72489.71121; time:1510542909.38
step 1596 of 3

step 1721 of 3000; error: 72295.66455; time:1510543443.40
step 1722 of 3000; error: 72294.24204; time:1510543447.63
step 1723 of 3000; error: 72292.82124; time:1510543451.88
step 1724 of 3000; error: 72291.40215; time:1510543456.12
step 1725 of 3000; error: 72289.98477; time:1510543460.34
step 1726 of 3000; error: 72288.56909; time:1510543464.57
step 1727 of 3000; error: 72287.15512; time:1510543468.77
step 1728 of 3000; error: 72285.74284; time:1510543473.00
step 1729 of 3000; error: 72284.33226; time:1510543477.20
step 1730 of 3000; error: 72282.92336; time:1510543481.42
step 1731 of 3000; error: 72281.51616; time:1510543485.60
step 1732 of 3000; error: 72280.11064; time:1510543489.81
step 1733 of 3000; error: 72278.70681; time:1510543494.00
step 1734 of 3000; error: 72277.30466; time:1510543498.23
step 1735 of 3000; error: 72275.90418; time:1510543502.44
step 1736 of 3000; error: 72274.50538; time:1510543506.66
step 1737 of 3000; error: 72273.10825; time:1510543510.85
step 1738 of 3

step 1863 of 3000; error: 72109.52446; time:1510544042.70
step 1864 of 3000; error: 72108.31827; time:1510544046.94
step 1865 of 3000; error: 72107.11343; time:1510544051.11
step 1866 of 3000; error: 72105.90994; time:1510544055.37
step 1867 of 3000; error: 72104.70780; time:1510544059.63
step 1868 of 3000; error: 72103.50701; time:1510544063.89
step 1869 of 3000; error: 72102.30756; time:1510544068.13
step 1870 of 3000; error: 72101.10945; time:1510544072.35
step 1871 of 3000; error: 72099.91268; time:1510544076.57
step 1872 of 3000; error: 72098.71725; time:1510544080.79
step 1873 of 3000; error: 72097.52315; time:1510544084.99
step 1874 of 3000; error: 72096.33039; time:1510544089.23
step 1875 of 3000; error: 72095.13896; time:1510544093.41
step 1876 of 3000; error: 72093.94885; time:1510544097.63
step 1877 of 3000; error: 72092.76007; time:1510544101.86
step 1878 of 3000; error: 72091.57262; time:1510544106.12
step 1879 of 3000; error: 72090.38649; time:1510544110.33
step 1880 of 3

step 2005 of 3000; error: 71950.86472; time:1510544642.52
step 2006 of 3000; error: 71949.83142; time:1510544646.74
step 2007 of 3000; error: 71948.79922; time:1510544650.95
step 2008 of 3000; error: 71947.76811; time:1510544655.19
step 2009 of 3000; error: 71946.73809; time:1510544659.46
step 2010 of 3000; error: 71945.70917; time:1510544663.71
step 2011 of 3000; error: 71944.68133; time:1510544667.90
step 2012 of 3000; error: 71943.65459; time:1510544672.12
step 2013 of 3000; error: 71942.62892; time:1510544676.32
step 2014 of 3000; error: 71941.60435; time:1510544680.56
step 2015 of 3000; error: 71940.58086; time:1510544684.76
step 2016 of 3000; error: 71939.55844; time:1510544689.01
step 2017 of 3000; error: 71938.53712; time:1510544693.21
step 2018 of 3000; error: 71937.51686; time:1510544697.45
step 2019 of 3000; error: 71936.49769; time:1510544701.67
step 2020 of 3000; error: 71935.47959; time:1510544705.92
step 2021 of 3000; error: 71934.46257; time:1510544710.13
step 2022 of 3

step 2147 of 3000; error: 71814.44571; time:1510545243.10
step 2148 of 3000; error: 71813.55421; time:1510545247.34
step 2149 of 3000; error: 71812.66362; time:1510545251.55
step 2150 of 3000; error: 71811.77394; time:1510545255.83
step 2151 of 3000; error: 71810.88516; time:1510545260.05
step 2152 of 3000; error: 71809.99729; time:1510545264.29
step 2153 of 3000; error: 71809.11033; time:1510545268.49
step 2154 of 3000; error: 71808.22427; time:1510545272.73
step 2155 of 3000; error: 71807.33911; time:1510545276.94
step 2156 of 3000; error: 71806.45485; time:1510545281.19
step 2157 of 3000; error: 71805.57150; time:1510545285.37
step 2158 of 3000; error: 71804.68904; time:1510545289.61
step 2159 of 3000; error: 71803.80748; time:1510545293.81
step 2160 of 3000; error: 71802.92682; time:1510545298.06
step 2161 of 3000; error: 71802.04705; time:1510545302.26
step 2162 of 3000; error: 71801.16818; time:1510545306.49
step 2163 of 3000; error: 71800.29020; time:1510545310.71
step 2164 of 3

step 2289 of 3000; error: 71696.45574; time:1510545842.78
step 2290 of 3000; error: 71695.68284; time:1510545847.01
step 2291 of 3000; error: 71694.91071; time:1510545851.21
step 2292 of 3000; error: 71694.13934; time:1510545855.45
step 2293 of 3000; error: 71693.36873; time:1510545859.71
step 2294 of 3000; error: 71692.59888; time:1510545863.95
step 2295 of 3000; error: 71691.82980; time:1510545868.19
step 2296 of 3000; error: 71691.06148; time:1510545872.42
step 2297 of 3000; error: 71690.29391; time:1510545876.61
step 2298 of 3000; error: 71689.52711; time:1510545880.88
step 2299 of 3000; error: 71688.76106; time:1510545885.11
step 2300 of 3000; error: 71687.99578; time:1510545889.34
step 2301 of 3000; error: 71687.23124; time:1510545893.54
step 2302 of 3000; error: 71686.46747; time:1510545897.81
step 2303 of 3000; error: 71685.70444; time:1510545902.05
step 2304 of 3000; error: 71684.94217; time:1510545906.28
step 2305 of 3000; error: 71684.18066; time:1510545910.47
step 2306 of 3

step 2431 of 3000; error: 71593.95569; time:1510546443.76
step 2432 of 3000; error: 71593.28281; time:1510546448.02
step 2433 of 3000; error: 71592.61057; time:1510546452.23
step 2434 of 3000; error: 71591.93898; time:1510546456.49
step 2435 of 3000; error: 71591.26803; time:1510546460.76
step 2436 of 3000; error: 71590.59772; time:1510546465.00
step 2437 of 3000; error: 71589.92806; time:1510546469.22
step 2438 of 3000; error: 71589.25903; time:1510546473.45
step 2439 of 3000; error: 71588.59065; time:1510546477.65
step 2440 of 3000; error: 71587.92291; time:1510546481.94
step 2441 of 3000; error: 71587.25581; time:1510546486.14
step 2442 of 3000; error: 71586.58935; time:1510546490.35
step 2443 of 3000; error: 71585.92353; time:1510546494.56
step 2444 of 3000; error: 71585.25834; time:1510546498.81
step 2445 of 3000; error: 71584.59379; time:1510546503.05
step 2446 of 3000; error: 71583.92987; time:1510546507.29
step 2447 of 3000; error: 71583.26659; time:1510546511.56
step 2448 of 3

step 2573 of 3000; error: 71504.52027; time:1510547045.47
step 2574 of 3000; error: 71503.93166; time:1510547049.72
step 2575 of 3000; error: 71503.34360; time:1510547054.00
step 2576 of 3000; error: 71502.75608; time:1510547058.31
step 2577 of 3000; error: 71502.16910; time:1510547062.52
step 2578 of 3000; error: 71501.58266; time:1510547066.77
step 2579 of 3000; error: 71500.99676; time:1510547071.00
step 2580 of 3000; error: 71500.41141; time:1510547075.24
step 2581 of 3000; error: 71499.82659; time:1510547079.45
step 2582 of 3000; error: 71499.24231; time:1510547083.71
step 2583 of 3000; error: 71498.65856; time:1510547087.95
step 2584 of 3000; error: 71498.07536; time:1510547092.19
step 2585 of 3000; error: 71497.49269; time:1510547096.39
step 2586 of 3000; error: 71496.91055; time:1510547100.65
step 2587 of 3000; error: 71496.32896; time:1510547104.87
step 2588 of 3000; error: 71495.74789; time:1510547109.17
step 2589 of 3000; error: 71495.16736; time:1510547113.44
step 2590 of 3

step 2715 of 3000; error: 71426.07329; time:1510547648.53
step 2716 of 3000; error: 71425.55542; time:1510547652.80
step 2717 of 3000; error: 71425.03801; time:1510547657.07
step 2718 of 3000; error: 71424.52105; time:1510547661.31
step 2719 of 3000; error: 71424.00455; time:1510547665.54
step 2720 of 3000; error: 71423.48850; time:1510547669.78
step 2721 of 3000; error: 71422.97290; time:1510547674.01
step 2722 of 3000; error: 71422.45775; time:1510547678.23
step 2723 of 3000; error: 71421.94306; time:1510547682.48
step 2724 of 3000; error: 71421.42881; time:1510547686.72
step 2725 of 3000; error: 71420.91502; time:1510547690.92
step 2726 of 3000; error: 71420.40168; time:1510547695.17
step 2727 of 3000; error: 71419.88878; time:1510547699.38
step 2728 of 3000; error: 71419.37634; time:1510547703.62
step 2729 of 3000; error: 71418.86434; time:1510547707.84
step 2730 of 3000; error: 71418.35279; time:1510547712.12
step 2731 of 3000; error: 71417.84169; time:1510547716.31
step 2732 of 3

step 2857 of 3000; error: 71356.83546; time:1510548250.33
step 2858 of 3000; error: 71356.37680; time:1510548254.57
step 2859 of 3000; error: 71355.91853; time:1510548258.89
step 2860 of 3000; error: 71355.46063; time:1510548263.15
step 2861 of 3000; error: 71355.00312; time:1510548267.36
step 2862 of 3000; error: 71354.54598; time:1510548271.61
step 2863 of 3000; error: 71354.08923; time:1510548275.80
step 2864 of 3000; error: 71353.63285; time:1510548280.02
step 2865 of 3000; error: 71353.17685; time:1510548284.26
step 2866 of 3000; error: 71352.72122; time:1510548288.50
step 2867 of 3000; error: 71352.26598; time:1510548292.74
step 2868 of 3000; error: 71351.81111; time:1510548296.98
step 2869 of 3000; error: 71351.35661; time:1510548301.20
step 2870 of 3000; error: 71350.90249; time:1510548305.45
step 2871 of 3000; error: 71350.44875; time:1510548309.67
step 2872 of 3000; error: 71349.99538; time:1510548313.90
step 2873 of 3000; error: 71349.54238; time:1510548318.10
step 2874 of 3

step 2999 of 3000; error: 71295.30414; time:1510548852.99
step 3000 of 3000; error: 71294.89504; time:1510548857.28
done in 13122.586s.


In [549]:
Preds=np.dot(fP,fQ.T)
Preds.shape

(943, 1682)

In [551]:
rating_pivot.columns.values

array([   1,    2,    3, ..., 1680, 1681, 1682])

In [552]:
rating_pivot.index.values

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [575]:
rawRating=[]
labels=[]
for id in array_test:
    #print id
    movie_id=id[1] #movie
    user_id=id[0] # user
    prediction=pd.DataFrame(Preds)
    r=prediction[movie_id-1][user_id-1]
    rawRating.append(r)
    labels.append(round(r))
    

In [565]:
labels[:10]

[4.0, 3.0, 4.0, 3.0, 2.0, 4.0, 5.0, 4.0, 4.0, 4.0]

In [566]:
print completeness_score(classes,labels)

0.243055407939


In [568]:
print homogeneity_score(classes,labels)

0.203494139909


In [583]:
err=0.0
for i in range(9430):
    error=abs(classes[i]-labels[i])
    err=err+error
print 'The overall mean absolute error(MAE) of prediction is: '
print err/9430

The overall mean absolute error(MAE) of prediction is: 
0.217511488158
